### PR TITLE
Procedurally generate `raw.rs` 

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a script that uses the `witx` crate to parse the
official WASI specification definitions and generate Rust bindings for
the upstream APIs. The intention is that this isn't actually a breaking
change in general (but it is in a few places that switch from `*mut
c_void` to `*mut u8`).

A CI check is also added to ensure that the in-repository `raw.rs` is
the same as what should be generated, which allows shipping this to
crates.io without actually parsing `*.witx` at build time.

### CODE REVIEW

**Comment by alexcrichton** [comment]
After this @sunfishcode told me about https://github.com/CraneStation/wasi-common/pull/152 which is the same spirit. I'm not sure if we want to share more infrastructure than we're already sharing (the `witx` crate), though? There's some parts that are for sure reusable across the two but they're largely not too too complicated so it may not be the worst thing to have some duplication for now

File: src/wasi_unstable/mod.rs [hunk 1]
```diff
@@ -559,7 +557,10 @@ pub unsafe fn poll_oneoff(in_: &[Subscription], out: &mut [Event]) -> Result<usi
 
 #[inline]
 pub fn proc_exit(rval: ExitCode) -> ! {
-    unsafe { __wasi_proc_exit(rval) }
+    unsafe {
+        __wasi_proc_exit(rval);
+        core::hint::unreachable_unchecked();
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/wasi_unstable/mod.rs]: 👍

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[submodule "crates/generate-raw/wasi"]
+	path = crates/generate-raw/wasi
+	url = https://github.com/webassembly/wasi
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply .gitmodules]: 👍

**Comment by alexcrichton** [comment]
That all sounds good to me, should be updated and green now!

**Comment by sunfishcode** [comment]
Looking at this more, I agree that this PR makes sense. This crate is a dependency of libc, which is a dependency of a lot of things, so having a checked-in raw.rs instead of parsing the witx at build time sounds desirable. And there's not all that much duplication in the end. So let's do it!

**Pull request merged**